A36-xds-for-servers.md

```diff
@@ -0,0 +1,316 @@
+ports are added after the server is created, the returned `grpc.Server` object
+may be an xds-aware `Server` to coordinate with the C API when `server.start()`
+is called. However, its implementation should be trivial as it could mostly
+delegate to a normal `server` instance.
+
+It would be possible to manage the lifetime of the XdsClient from the `server`,
+based on `server.stop()`, but given C++ lacks this avenue the C API probably
+will not need this notification.
+
+
+### Java
+
+Create an `XdsServerBuilder` that extends `ServerBuilder` and delegates to a
+"real" builder. The server returned from `builder.build()` would be an xDS-aware
+server delegating to a "real" server instance. The `XdsServerBuilder` would
+install an `AtomicReference&lt;ServerInterceptor>`-backed interceptor in the
+built server and pass the listening port and interceptor reference to the
+xDS-aware server when constructed. Since interceptors cannot be removed once
+installed, the builder may only be used once (`build()` should throw if called a
+second time).
+
+The xDS-aware server’s `start()` will create the `XdsClient` with the passed
+port and wait for initial configuration before delegating to the real `start()`.
+If the xDS bootstrap is missing it will throw an IOException within `start()`.
+The `XdsClient` will be shut down on `shutdownNow()`/termination. When new or
+updated configuration is received, it will create a new intercepter and update
+the interceptor reference.
+
+XdsServerBuilder will have an `xdsInitListener(XdsInitListener)`.
+`XdsInitListener` will have the one method `xdsLoadFailure(IOException)` that is
+called for failures loading the initial configuration within `start()`, since
+`start()` itself does not fail in that case. A default implementation will log
+the exception. If a user sets the listener, it will replace the default logging
+implementation.
+
+In order to allow transport-specific configuration, the `XdsServerBuilder` will
+have a `@ExperimentalApi ServerBuilder transportBuilder()` method whose return
+value can be cast to `NettyServerBuilder` for experimental API configuration.
+
+`NettyServerBuilder` will add an internal method `eagAttributes(Attributes)`. It
+will plumb those attributes to `GrpcHttp2ConnectionHandler.getEagAttributes()`
+as implemented by `NettyServerHandler`, which currently just returns
+`Attributes.EMPTY`. `XdsServerBuilder` will then specify `eagAttributes` to
+inject credential information for the `XdsServerCredential`, similar to
+client-side, although with the added step of needing to process the
+FilterChainMatch for the specific connection.
+
+
+### Go
+
+Create an `xds.Server` struct that would internally contain a `grpc.Server`. It
+would inject its own `StreamServerInterceptor` and `UnaryServerInterceptor`s
+into the `grpc.Server`. The interceptors would be controlled by an
+`atomic.Value` that would update with the configuration. When
```

Comment:
Could we skip this line about using an `atomic` to update the interceptors? It seems very implementation specific and could change anytime. Currently, we do have an interceptor which does nothing, and is not changed based on any configuration received. When we start supporting changing of the interceptors based on received configuration, we could use an `atomic` or a `mutex`.